Update replicator configuration and implementation

jayant-dhingra-cb opened the pull request.

jayant-dhingra-cb committed:
Update replicator configuration and implementation

velicuvlad commented on cblite/src/collection-config.ts:
```diff
@@ -1,88 +1,130 @@
-  constructor(
-    channels: string[] | null | undefined,
-    documentIds: string[] | null | undefined
-  ) {
-    this.channels = channels ?? [];
-    this.documentIds = documentIds ?? [];
+  constructor() {
+    this.channels = [];
+    this.documentIDs = [];
+    this.pullFilter = undefined;
+    this.pushFilter = undefined;
   }
```
I think default empty string would suite better as it will point that string injection is done for the filters onto CBL side.

velicuvlad commented on cblite/src/replicator-configuration.ts:
```diff
@@ -72,41 +136,173 @@ export class ReplicatorConfiguration {
+   */
+  addCollection(collection: Collection, config: CollectionConfig): void {
+    if (this.isNewApi) {
+      throw new Error('Cannot call addCollection() on NEW API instance. Collections must be provided at construction.');
```
Interesting, we didn't guard against this on CBL. I don't know what happens when creating a ReplicatorConfig from [CollectionConfiguration[ and then adding another one via `.addCollection` nor what the expectation is. 

This should be pointed out in the RN docs and brought to PM/Adam, if not already, as in theory CBL allows for this.

jayant-dhingra-cb replied:
So, when a person is using the older way, they can use the addCollection method. But if the person is using the newer way that is the new API, then they cannot add the collection. It will throw the error.

velicuvlad commented on cblite/src/replicator-configuration.ts:
```diff
@@ -72,41 +136,173 @@ export class ReplicatorConfiguration {
-      config = new CollectionConfig(undefined, undefined);
+  addCollections(collections: Collection[], config: CollectionConfig): void {
+    if (this.isNewApi) {
+      throw new Error('Cannot call addCollections() on NEW API instance. Collections must be provided at construction.');
```
same here

jayant-dhingra-cb replied:
SAME - So, when a person is using the older way, they can use the addCollection method. But if the person is using the newer way that is the new API, then they cannot add the collection. It will throw the error.

velicuvlad commented on cblite/src/replicator-configuration.ts:
```diff
@@ -72,41 +136,173 @@ export class ReplicatorConfiguration {
+   */
+  removeCollection(collection: Collection): void {
+    if (this.isNewApi) {
+      throw new Error('Cannot call removeCollection() on NEW API instance. Collections are immutable.');
```
I have not tested this either in a mix API call for CBL.

jayant-dhingra-cb replied:
We had to do it since javascript does not allow function overloading and we need the same function name for both old api and new api

velicuvlad commented on cblite/src/replicator-configuration.ts:
```diff
@@ -72,41 +136,173 @@ export class ReplicatorConfiguration {
+    for (const [collections, _] of this.collectionsMap.entries()) {
+      if (collections.includes(collection)) {
+        const filtered = collections.filter(c => c !== collection);
+        if (filtered.length === 0) {
```
this makes sure that collections is dropped when the last collection is dropped as well. 

It should also delete it from the Map when is not the last one. I might not understand how filter works, does it do that already?

velicuvlad commented on cblite/src/replicator-configuration.ts:
```diff
@@ -489,111 +631,288 @@ export class ReplicatorConfiguration {
+      );
+      config.collectionConfig = JSON.stringify(collectionConfigArray);
+      console.log('[ReplicatorConfiguration.toJson] ✅ NEW API serialization complete');
     } else {
```
this check exists on CBL level. not needed here, disregard

velicuvlad commented on cblite/src/replicator-configuration.ts:
```diff
@@ -489,111 +631,288 @@ export class ReplicatorConfiguration {
+   * 
+   * @private
+   */
+  private validateCollectionsScopeAndDatabase(): boolean {
```
this can be removed as it exists at CBL level so it will throw from iOS directly.

jayant-dhingra-cb replied:
I have to check whether we are properly, propagating the error or not from native side, that's why we have it.

jayant-dhingra-cb committed:
Update collection config and replicator configuration

jayant-dhingra-cb committed:
Merge remote-tracking branch 'origin/RN-v1' into new_replicator_config

jayant-dhingra-cb committed:
refactor: update collectionsMap type and remove debug logs

- Changed collectionsMap to allow undefined values in CollectionConfig.
- Removed console log statements from clone and toJson methods for cleaner output.
